### PR TITLE
Propagate NaN in y for scipy.special.xlogy/xlog1py at x=0

### LINKED ISSUE #37751: `jax.scipy.special.xlogy` and `xlog1py` return `0.0` instead of `nan` when `x=0` and `y=nan`
### Description

`jax.scipy.special.xlogy(x, y)` and `jax.scipy.special.xlog1py(x, y)` are supposed to compute `x * log(y)` and `x * log(1+y)` respectively, with the special case that the result is `0` when `x == 0` (to handle `0 * log(0)` gracefully).

However, when `x=0` and `y=nan`, JAX returns `0.0` instead of `nan`. This is incorrect — a `nan` input should propagate to the output because the mathematical expression `0 * log(nan)` is indeterminate. `scipy.special.xlogy` correctly returns `nan` in this case.

This breaks any downstream computation that depends on NaN propagation to detect invalid inputs — for example, log-likelihood computations in probabilistic models where `nan` in parameters should surface as `nan` in the loss.

## Minimal reproducer

```python
import jax.numpy as jnp
from jax.scipy import special as lsp_special
import scipy.special as osp_special
import numpy as np

x = jnp.float32(0.0)
y = jnp.float32(float("nan"))

print(osp_special.xlogy(0.0, float("nan")))   # nan  ← correct
print(lsp_special.xlogy(x, y))                 # 0.0  ← BUG: should be nan

print(osp_special.xlog1py(0.0, float("nan")))  # nan  ← correct
print(lsp_special.xlog1py(x, y))               # 0.0  ← BUG: should be nan
```

## Expected behavior

```
nan
nan
nan
nan
```

## Actual behavior

```
nan
0.0
nan
0.0
```

## Root cause

The JAX implementation short-circuits to `0` whenever `x == 0`, without first checking whether `y` is `nan`. The correct behavior is to return `nan` if either argument is `nan`, and only return `0` for `x == 0` when `y` is a valid (non-NaN) value. `scipy.special.xlogy` follows this contract.

### System info (python version, jaxlib version, accelerator, etc.)

```
jax:     0.10.1.dev20260513+64f6e229b
jaxlib:  0.10.0
numpy:   2.4.4
scipy:   1.17.1
python:  3.12.3
OS:      Linux-6.14.0-37-generic-x86_64-with-glibc2.39
backend: cpu
```

### PR DESCRIPTION
**Pull request by adityasingh2400**
## Summary

`jax.scipy.special.xlogy(0, nan)` and `xlog1py(0, nan)` currently return `0.0`, but `scipy.special.xlogy(0, nan)` returns `nan`. This breaks NaN propagation in downstream computations such as log-likelihoods, where a `nan` parameter should surface as `nan` in the loss.

The implementation in `jax/_src/scipy/special.py` short-circuits to `0` whenever `x == 0`, without first checking whether `y` is `nan`. This PR restores SciPy parity by checking `isnan(y)` separately and propagating NaN even on the `x == 0` short-circuit.

## Reproducer (from #37751)

```python
import jax.numpy as jnp
from jax.scipy import special as lsp_special
import scipy.special as osp_special

print(osp_special.xlogy(0.0, float("nan")))    # nan
print(lsp_special.xlogy(jnp.float32(0.0), jnp.float32(float("nan"))))  # was 0.0, now nan

print(osp_special.xlog1py(0.0, float("nan")))  # nan
print(lsp_special.xlog1py(jnp.float32(0.0), jnp.float32(float("nan"))))  # was 0.0, now nan
```

## Changes

- `jax/_src/scipy/special.py`: in `xlogy` and `xlog1py`, wrap the existing `where(x_ok, ...)` result in a second `where(isnan(y), y, result)` so NaN in `y` is preserved at `x == 0`.
- `tests/lax_scipy_test.py`: add `testXlogyPropagatesNaNInY` and `testXlog1pyPropagatesNaNInY` covering NaN, finite-zero, and a typical value.

The existing `testXlogyShouldReturnZero`, `testXlog1pyShouldReturnZero`, and the gradient tests at `x == 0` remain unaffected since they only exercise finite `y`.

## Test plan

- [x] `xlogy(0, nan)` returns `nan`, `xlogy(0, 1)` returns `0`, `xlogy(1, 2)` returns `log(2)`
- [x] Same three cases for `xlog1py`
- [x] Existing `xlogy(0, 0) == 0` and `xlog1py(0, -1) == 0` still pass
- [x] Gradient at `(0, y)` for finite `y` unchanged (custom JVP path is untouched)

Fixes #37751

### CODE REVIEW

**Comment by adityasingh2400** [comment]
@googlebot I signed it!

**Comment by adityasingh2400** [comment]
Applied the single-jnp.where simplification in b85fff0. The new form relies on `0 * NaN = NaN` from IEEE 754 to propagate NaN through the multiplication, matching the prior two-step semantics. Verified parity with scipy across (x=0, y=NaN), (x=0, y=finite), (x!=0, y=NaN), (x!=0, y=0), and (x!=0, y<0). Test coverage unchanged.

File: tests/lax_scipy_test.py [hunk 1]
```diff
@@ -270,6 +270,22 @@ def testGradOfXlogyAtZero(self):
   def testXlog1pyShouldReturnZero(self):
     self.assertAllClose(lsp_special.xlog1py(0., -1.), 0., check_dtypes=False)
 
+  def testXlogyPropagatesNaNInY(self):
+    # https://github.com/jax-ml/jax/issues/37751
+    # scipy.special.xlogy(0, nan) returns nan; ensure JAX matches.
+    self.assertTrue(np.isnan(lsp_special.xlogy(0., float("nan"))))
+    self.assertAllClose(lsp_special.xlogy(0., 1.), 0., check_dtypes=False)
```
Comment (jakevdp) [inline tests/lax_scipy_test.py]: Lines 277-279 seem superfluous, remove?

File: tests/lax_scipy_test.py [hunk 2]
```diff
@@ -270,6 +270,22 @@ def testGradOfXlogyAtZero(self):
...
+    self.assertAllClose(lsp_special.xlogy(1., 2.), np.log(2.),
+                        check_dtypes=False)
+
+  def testXlog1pyPropagatesNaNInY(self):
+    # https://github.com/jax-ml/jax/issues/37751
+    # scipy.special.xlog1py(0, nan) returns nan; ensure JAX matches.
+    self.assertTrue(np.isnan(lsp_special.xlog1py(0., float("nan"))))
+    self.assertAllClose(lsp_special.xlog1py(0., 1.), 0., check_dtypes=False)
```
Comment (jakevdp) [inline tests/lax_scipy_test.py]: Same with lines 285-287.

File: tests/lax_scipy_test.py [hunk 3]
```diff
@@ -270,6 +270,22 @@ def testGradOfXlogyAtZero(self):
   def testXlog1pyShouldReturnZero(self):
     self.assertAllClose(lsp_special.xlog1py(0., -1.), 0., check_dtypes=False)
 
+  def testXlogyPropagatesNaNInY(self):
+    # https://github.com/jax-ml/jax/issues/37751
+    # scipy.special.xlogy(0, nan) returns nan; ensure JAX matches.
+    self.assertTrue(np.isnan(lsp_special.xlogy(0., float("nan"))))
```
Comment (jakevdp) [inline tests/lax_scipy_test.py]: Use `self.assertAllClose(lsp_special.xlogy(0., np.nan), np.nan)` – `assertAllClose` has NaN check logic in it, and on failure the error will be more informative.

File: tests/lax_scipy_test.py [hunk 4]
```diff
@@ -270,6 +270,22 @@ def testGradOfXlogyAtZero(self):
...
+    self.assertAllClose(lsp_special.xlogy(0., 1.), 0., check_dtypes=False)
+    self.assertAllClose(lsp_special.xlogy(1., 2.), np.log(2.),
+                        check_dtypes=False)
+
+  def testXlog1pyPropagatesNaNInY(self):
+    # https://github.com/jax-ml/jax/issues/37751
+    # scipy.special.xlog1py(0, nan) returns nan; ensure JAX matches.
+    self.assertTrue(np.isnan(lsp_special.xlog1py(0., float("nan"))))
```
Comment (jakevdp) [inline tests/lax_scipy_test.py]: `assertAllClose` here as well, please

**Review by jakevdp — CHANGES_REQUESTED** [review]
Thanks! A couple comments below.

**Comment by adityasingh2400** [comment]
Switched to assertAllClose in 24c2321, both places.

**Comment by adityasingh2400** [comment]
Dropped lines 277-279 and 285-287 in 14396d3. Both calls were already covered by testXlogyShouldReturnZero, testXlog1pyShouldReturnZero, and the lax_scipy_special_functions parity sweep, so the NaN tests are now single-assertion.

**Review by jakevdp — CHANGES_REQUESTED** [review]
Looks good, thanks! Please squash your changes into a single commit and we'll get this merged!

**Comment by adityasingh2400** [comment]
Squashed into a single commit (4842430). Diff is unchanged from the reviewed state, so this is ready to merge whenever you are. Thanks!

**Comment by jakevdp** [comment]
Something went wrong with the squash: I suspect you rebased on an outdated main branch.

**Comment by adityasingh2400** [comment]
You are right, good catch. My earlier squash kept the tree from a stale main while setting the parent to a newer main, so the diff picked up a reversal of everything that had landed in between.

Fixed in 60912d9. I replayed the three original commits onto current main and squashed them there, so the PR is now just the two files it should be: the `x_ok | jnp.isnan(y)` change in `jax/_src/scipy/special.py` and the two propagation tests in `tests/lax_scipy_test.py`. Sorry for the noise.

**Review by jakevdp — APPROVED** [review]
Thanks!
